### ISSUE TITLE
Unsupported Tensor Layout: Sparse

### ISSUE THREAD

**Issue by bytesnake** [LaurentMazare/tch-rs#131]
Hello,
first thanks for the project! I'm trying to train a word embedding with a single layer NN. After implementing the `embedding_bag` type and setting sparsity to `true`, I get the following error. 
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TorchError { c_error: "unsupported tensor layout: Sparse (validate at ../../aten/src/ATen/native/TensorIterator.h:122)\nframe 
#0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6a (0x7f1f5bb16c4a in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libc10.so)\nframe 
#1: <unknown function> + 0xb5d3f6 (0x7f1f5c88e3f6 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#2: at::TensorIterator::binary_op(at::Tensor&, at::Tensor const&, at::Tensor const&, bool) + 0xe6 (0x7f1f5cb07916 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#3: at::native::mul(at::Tensor const&, at::Tensor const&) + 0x48 (0x7f1f5c853608 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#4: at::native::mul(at::Tensor const&, c10::Scalar) + 0x4b (0x7f1f5c8554ab in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#5: <unknown function> + 0x1148a86 (0x7f1f5ce79a86 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#6: <unknown function> + 0x101d9a6 (0x7f1f5cd4e9a6 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#7: <unknown function> + 0xbb7be7 (0x7f1f5c8e8be7 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#8: <unknown function> + 0x2f5a56c (0x7f1f5ec8b56c in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#9: <unknown function> + 0x101d9a6 (0x7f1f5cd4e9a6 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#10: at::Tensor::mul(c10::Scalar) const + 0x175 (0x7f1f5c933b95 in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#11: torch::optim::SGD::step() + 0x68d (0x7f1f5f0d136d in target/release/build/torch-sys-b700df52b668235a/out/libtorch/libtorch/lib/libtorch.so)\nframe 
#12: <unknown function> + 0x3db07 (0x5645b05d3b07 in target/release/sample)\nframe 
__libc_start_main + 0xf3 (0x7f1f5b8c8153 in /usr/lib/libc.so.6)\nframe 
...
#21: <unknown function> + 0x174ae (0x5645b05ad4ae in target/release/sample)\n" }
```
I already switched from `Adam` to `SGD` (which should support sparse model optimization), but the update step seems to fail ..? The backpropagation step works fine.
Some questions:
1. Do we actually support sparse matrices?
2. How can I implement a custom optimizer? I already tried to get the gradients from my loss function, but it just returns an empty Tensor
3. There seems to be no `SparseAdam` in the pytorch C++ API, but its supported in python, is there any way to import that into the project?
4. Could we improve the error output, at least split at `\nframe` for the output :D

**Comment by LaurentMazare** [comment]
A good first thing to try if you haven't already is setting `sparse` to `false` and check that everything works fine here.
This crates mostly relies on auto-generated bindings using the C++ api so the level of coverage for sparse matrices should be the same as the C++ api. I imagine that basic operations are well supported but as you noticed there is no `SparseAdam`. You could either write a C++ version for it by adapting the Python version and the current [Adam C++ implementation](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/optim/adam.cpp). 
Another possibility would be to write a pure rust version based on the Python implementation. The Python implementation uses `is_sparse` and `coalesce` to extract the indices to update, I don't think `is_sparse` is already exposed, I'll add it if that's not the case. `coalesce` should already be available.

**Comment by bytesnake** [comment]
thank you! I will look into `coalesce` tomorrow. I obviously started with `sparse` as `false`, but the backprop is just too slow for a large number of ngrams ..

**Comment by LaurentMazare** [comment]
Note that the `is_sparse` method should now be available in case it helps.

**Comment by bytesnake** [comment]
I'm using now Adam for dense layer and SGD for the embedding matrix. An implementation for Adam (for both) is [here](https://gist.github.com/bytesnake/f31c787f8b3bbad59dd82f31b75c4078) I tested the optimizer on the MNIST dataset.

 * The python version does not support normal Adam for dense gradients, is there some reason to do that?
 * We don't support any method for sparse tensor addition, because of that only matrices are supported. I think the `spadd` function is missing, but the documentation of pytorch C++ API is really bad.

**Comment by bytesnake** [comment]
Added an example with custom Adam optimizer in #135

The `spadd` function can be replicated with a combination of `index_select_` and `index_add_` :)
